torch/csrc/jit/fusion_compiler.cpp

```diff
@@ -1047,67 +1132,407 @@ struct CPUFusionFunction : public CompiledFusionFunction {
+      }
+    } else {
+      auto group_map_size = getMapSize(args, broadcast_group);
+      // NB: this checks that group_map_size is defined AND equal to map_size
+      if (map_size != group_map_size) {
+        return at::nullopt;
+      }
+    }
   }
-  for(auto & output : graph.outputs()) {
-    auto t = output->type()->expect<TensorType>();
-    agraph.output_desc.emplace_back(t);
+  return map_size;
+}
+
+void FusedKernelCache::runFallback(Stack& stack) {
+  InterpreterState(fallback_code).runOneStage(stack);
+}
+
+// NB: args are mutated in this call. map_size is mutated too, but is restored to its original
+// value before this function returns (it's an optimization).
+void FusedKernelCache::expandArgs(std::vector<at::Tensor>& args, std::vector<int64_t>& map_size) {
+  for (size_t i = 0; i < args.size(); ++i) {
+    auto & arg = args[i];
+    auto & pdesc = input_chunks[i];
+    if (pdesc.nSubtensors == 1) {
+      if (arg.sizes().equals(map_size)) continue;
+      arg = arg.expand(map_size);
+    } else {
+      map_size.at(pdesc.dim) *= pdesc.nSubtensors;
+      if (!arg.sizes().equals(map_size)) {
+        arg = arg.expand(map_size);
+      }
+      map_size.at(pdesc.dim) /= pdesc.nSubtensors;
+    }
   }
-  return getOrCompile(agraph);
 }
 
+std::unique_ptr<FusedKernel> FusedKernelCache::compileSpec(
+      const FusedKernelArgSpec& spec, const std::vector<int64_t>& map_size) {
+  AnnotatedGraph agraph {*graph, device};
 
-std::shared_ptr<CompiledFusionFunction> FusionCompiler::getOrCompile(Graph & graph,
-                                                     int device,
-                                                     at::ArrayRef<at::Tensor> inputs,
-                                                     at::ArrayRef<at::Tensor> outputs) {
-  AnnotatedGraph agraph(graph, device);
-  for(auto & i : inputs) {
-   agraph.input_desc.emplace_back(i);
+  agraph.input_desc = spec.descs();
+  at::optional<at::ScalarType> scalar_type;
+  for (TensorDesc& desc : agraph.input_desc) {
+    if (isFloatingType(desc.scalar_type)) {
+      scalar_type = desc.scalar_type;
+      break;
+    }
   }
-  for(auto & i : outputs) {
-   agraph.output_desc.emplace_back(i);
+  JIT_ASSERT(scalar_type);
+
+  for (Value * output : graph->outputs()) {
+    std::vector<int64_t> sizes = map_size;
+    if (output->node()->kind() == prim::FusedConcat) {
+      sizes.at(output->node()->i(attr::dim)) *= output->node()->inputs().size();
+    }
+    auto type = CompleteTensorType::create(*scalar_type, device, sizes);
+    agraph.output_desc.emplace_back(std::move(type));
   }
-  return getOrCompile(agraph);
+
+  std::string name = "kernel_" + std::to_string(next_kernel_id++);
+  FusedKernel * raw_func;
+  if (device != kCPUDevice) {
+#ifdef USE_CUDA
+    raw_func = new CUDAFusedKernel(name, agraph);
+#else
+    throw std::runtime_error("cannot compile a CUDA fusion group, CUDA is not enabled.");
+#endif
+  } else {
+    JIT_ASSERT(compiler.canCompileOnCPU());
```

Comment:
Before we release, we need to fix the fact that on Windows the CPU fuser doesn't run, and cause the GE to fail if it did a fusion. An easy workaround would have FusionCompiler just immediately go to fallback mode in this case. That way the decision whether we can do CPU fusion is localized to only the fusion compiler and not both the optimization pass and the compiler.

Reply:
You mean this as a followup task, correct?